apps/wallet/src/ui/app/shared/faucet/useFaucetRateLimiter.ts

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useEffect, useState } from 'react';
+
+const FAUCET_RATE_LIMIT_EXPIRY_TIME_KEY = 'faucet_rate_limit_expiry_time';
+
+// We'll rate limit users for 20 minutes which should
+// more-or-less mock how the faucet API rate limits users
+const rateLimitExpiryTime = 20 * 60 * 1000;
+
+export function useFaucetRateLimiter() {
+    const [isRateLimited, setRateLimited] = useState(() => {
+        const expiryTime = localStorage.getItem(
+            FAUCET_RATE_LIMIT_EXPIRY_TIME_KEY
+        );
+        const currTime = new Date().getTime();
+        return currTime <= Number(expiryTime);
+    });
+
+    const rateLimit = () => {
+        const currTime = new Date().getTime();
+        const expiryTime = currTime + rateLimitExpiryTime;
+
+        localStorage.setItem(
+            FAUCET_RATE_LIMIT_EXPIRY_TIME_KEY,
+            String(expiryTime)
+        );
+        setRateLimited(true);
+    };
+
+    useEffect(() => {
```

Comment:
I opted for an effect here since state initializers are supposed to be pure functions